### ISSUE TITLE
Component Library Proposal: Fields

### ISSUE THREAD

**Issue by spencer516** [Ticketfly/ticketfly-ui#153]
@Ticketfly/frontenders : Yesterday Brian, Melissa, and I discussed some ways we want to start pushing this project forward. First, we agreed that, even if they are in the same repository in the short term, we should think of CSS and Javascript as completely separate concerns. Brian has presented some ideas for how to architect the CSS part of this project (see those other open issues if you have anything to add). 

For the consumption of that CSS library through Ember Components, our next agreed step is to propose an "API" for each category of these components before we start implementing any of them. 

So, this issue outlines a proposed way of structuring all the components that can be composed together to implement all the features from the "Fields" section of the design working file. 

(This proposal does assume the existence of a `tf-icon` component — the merits/api of that will be in a separate issue.)

In order to implement this, we would create 3 top-level components (`tf-input`, `tf-input-group`, `tf-icon-input`) and two "not public" nested components (`tf-input/input-label`, `tf-input/input-description`). 

Feedback, questions, thoughts, and concerns all welcome!

**Input Sizing**

TEMPLATE:

```hbs
{{tf-input size="large"}}
{{tf-input}} {{!-- size="default" by default --}}
{{tf-input size="small"}}
```

RESULTING HTML:

```html
<input type="text" class="c-input u-large">
<input type="text" class="c-input">
<input type="text" class="c-input u-small">
```

**Input Block with Label & Description**

TEMPLATE:

```hbs
{{#tf-input-group size="large" as |group|}}
  {{group.label "First Name"}}
  {{group.input}}
  {{group.description "You must provide a First Name"}}
{{/tf-input-group}}
```

RESULTING HTML:

```html
<div class="c-input-group">
  <label class="u-large" for="random-uuid">First Name</label>
  <input class="c-input u-large" type="text" id="random-uuid" aria-describedby="other-uuid">
  <p class="u-large" id="other-uuid">You must provide a First Name</p>
</div>
```

**Input Validation States**

TEMPLATE:

```hbs
{{tf-input should-validate=true is-valid=true}}
{{tf-input should-validate=true is-valid=false}}
{{tf-input should-validate=false is-valid=true}}

{{#tf-input-group should-validate=true is-valid=false as |group|}}
  {{group.label "First Name"}}
  {{group.input}}
  {{group.description "You must provide a First Name"}}
{{/tf-input-group}}
```

RESULTING HTML:

```html
<input type="text" class="c-input is-valid">
<input type="text" class="c-input is-invalid">
<input type="text" class="c-input">

<div class="c-input-group">
  <label class="is-invalid" for="random-uuid">First Name</label>
  <input class="c-input is-invalid" type="text" id="random-uuid" aria-describedby="other-uuid">
  <p class="is-invalid" id="other-uuid">You must provide a First Name</p>
</div>
```

**Inputs with Icons**

TEMPLATE:

```hbs
{{#tf-icon-input as |i|}}
  {{i.icon "$"}}
  {{i.input}}
  {{i.icon (component "tf-icon" "down-arrow")}}
{{/tf-icon-input}}
```

RESULTING HTML (if the input has focus):

```html
<div class="c-icon-input-group is-focused">
  <span class="c-icon-input__icon">$</span>
  <input class="c-icon-input" type="text">
  <span class="c-icon-input__icon">
    <span class="c-tf-icon tf-calendar"></span>
  </span>
</div>
```

**Text Area**

TEMPLATE:

```hbs
{{tf-textarea}}
```

RESULTING HTML:

```html
<textarea class="c-textarea"></textarea>
```

**Date Input**

This component will be composed of an `tf-input-group` and a `tf-icon-input` components.
Might be implemented something like (with a bunch of other stuff going on in the
javascript component definition):

```hbs
{{#tf-input-group as |group|}}
  {{group.label "Date Input"}}
  {{#group.icon-input as |i|}}
    {{i.icon (component "if-icon" "calendar" is-active=(or date))}}
    {{i.input value=date update=(action "updateDate")}}
  {{/group.icon-input}}
{{/tf-input-group}}
```

**Comment by bl4ckm0r3** [comment]
i disagree on this

```js
{{#tf-input-group size="large" as |group|}}
  {{group.label "First Name"}}
  {{group.input}}
  {{group.description "You must provide a First Name"}}
{{/tf-input-group}}
```

grouping elements has nothing to do with defining their size ;)

**Comment by spencer516** [comment]
@bl4ckm0r3 — I don't have a strong opinion on this one. But, the thought was that this was just a way to setting the `size="large"` on the contextual components. This would avoid having to do this:

```hbs
{{#tf-input-group as |group|}}
  {{group.label "First Name" size="large"}}
  {{group.input size="large"}}
  {{group.description "You must provide a First Name" size="large"}}
{{/tf-input-group}}
```

I think it's right that it might be a bit misleading — but, it also feels pretty convenient. Maybe. Thoughts? Anyone else have an opinion on this?

**Comment by zhujy8833** [comment]
@spencer516 Does that support the case of overriding the size for child components?
such as 
```hbs
{{#tf-input-group size="large" as |group|}}
  {{group.label "First Name" }} <!--large, following group size-->
  {{group.input }}  <!--large, following group size-->
  {{group.description "You must provide a First Name" size="small"}} <!--small, overriding group size-->
{{/tf-input-group}}
```
I think it makes sense to define a group size on the top level, in case all of the children can just inherit it and don't need to repeat it. If specific child needs customization, they can define on its own level

**Comment by spencer516** [comment]
Yup! That would work. Under the hood, the `tf-input-group` template would look something like:

```hbs
{{yield (hash
  label=(component 'tf-input/input-label' size=size)
  input=(component 'tf-input' size=size)
  description=(component 'tf-input/input-description' size=size)
)}}
```

**Comment by BrianSipple** [comment]
Yeah, I've generally found that sizing is better off left as an outside concern. A component shouldn't have to know or care about that.

**Comment by spencer516** [comment]
@BrianSipple — The latter is the intent of that, indeed. 

But — your misreading it initially is something worth exploring further as this didn't pass your default sniff test...so, can you elaborate on what you thought it was vs what it is? Maybe the name isn't clear? Maybe we should have a convention for attributes that affect styles/class names versus attributes that are bound for javascripty things?

So, maybe something like this could be clearer where things bound for class names are prefixed with `styled-*`?

```hbs
{{#tf-input-group styled-size="large" as |group|}}
  {{group.label "First Name" }}
  {{group.input }}
  {{group.description "You must provide a First Name" styled-size="small"}}
{{/tf-input-group}}
```

(Also, do we have an opinion around dasherized vs camelized attribute names? I feel like this has come up; don't remember where we landed).

**Comment by BrianSipple** [comment]
@spencer516 From what I've seen, I believe it's camelCase

**Comment by ksin** [comment]
Initial implementation based on specs above https://github.com/Ticketfly-UI/ticketfly-fields

**Comment by spencer516** [comment]
Closing as we've implemented.
